lib/vtls/schannel.c

```diff
@@ -405,6 +405,362 @@ get_cert_location(TCHAR *path, DWORD *store_name, TCHAR **store_path,
+        filesize = ftell(fInCert);
+      if(filesize < 0)
+        continue_reading = 0;
+      if(continue_reading)
+        continue_reading = fseek(fInCert, 0, SEEK_SET) == 0;
+      if(continue_reading)
+        certdata = malloc(((size_t)filesize) + 1);
+      if(!certdata ||
+         ((int) fread(certdata, (size_t)filesize, 1, fInCert) != 1))
+        continue_reading = 0;
+      fclose(fInCert);
+      curlx_unicodefree(cert_path);
+
+      if(!continue_reading) {
+        failf(data, "schannel: Failed to read cert file %s",
+              data->set.ssl.primary.clientcert);
+        free(certdata);
+        return CURLE_SSL_CERTPROBLEM;
+      }
+
+      /* Convert key-pair data to the in-memory certificate store */
+      datablob.pbData = (BYTE*)certdata;
+      datablob.cbData = (DWORD)filesize;
+
+      if(data->set.ssl.key_passwd != NULL)
+        pwd_len = strlen(data->set.ssl.key_passwd);
+      pszPassword = (WCHAR*)malloc(sizeof(WCHAR)*(pwd_len + 1));
+      if(pwd_len > 0)
+        str_w_len =
+          MultiByteToWideChar(CP_UTF8, MB_ERR_INVALID_CHARS,
+                              data->set.ssl.key_passwd, (int)pwd_len,
+                              pszPassword, (int)(pwd_len + 1));
+
+      if((str_w_len >= 0) && (str_w_len <= (int)pwd_len))
+        pszPassword[str_w_len] = 0;
+      else
+        pszPassword[0] = 0;
+
+      cert_store = PFXImportCertStore(&datablob, pszPassword, 0);
+      free(pszPassword);
+      free(certdata);
+      if(!cert_store) {
+        DWORD errorcode = GetLastError();
+        if(errorcode == ERROR_INVALID_PASSWORD)
+          failf(data, "schannel: Failed to import cert file %s, "
+                "password is bad", data->set.ssl.primary.clientcert);
+        else
+          failf(data, "schannel: Failed to import cert file %s, "
+                "last error is 0x%x", data->set.ssl.primary.clientcert,
+                errorcode);
+        return CURLE_SSL_CERTPROBLEM;
+      }
+
+      client_certs[0] = CertFindCertificateInStore(
+          cert_store, X509_ASN_ENCODING | PKCS_7_ASN_ENCODING, 0,
+          CERT_FIND_ANY, NULL, NULL);
+
+      if(!client_certs[0]) {
+        failf(data, "schannel: Failed to get certificate from file %s"
+              ", last error is 0x%x",
+              data->set.ssl.primary.clientcert, GetLastError());
+        CertCloseStore(cert_store, 0);
+        return CURLE_SSL_CERTPROBLEM;
+      }
+
+      schannel_cred.cCreds = 1;
+      schannel_cred.paCred = client_certs;
+    }
+    else {
+      cert_store =
+        CertOpenStore(CURL_CERT_STORE_PROV_SYSTEM, 0,
+                      (HCRYPTPROV)NULL,
+                      CERT_STORE_OPEN_EXISTING_FLAG | cert_store_name,
+                      cert_store_path);
+      if(!cert_store) {
+        failf(data, "schannel: Failed to open cert store %x %s, "
+              "last error is 0x%x",
+              cert_store_name, cert_store_path, GetLastError());
+        free(cert_store_path);
```

Comment:
Again overwriting it with zeros wouldn't hurt much I guess even though this is probably not as critical as the password

Reply:
Existing code. Just pulled out into its own method. We can refactor later. Let's not distract ourselves.